File: lab_5/generators/random_generator.cc
#include <iostream>
#include <fstream>
#include <bitset>
#include <random>
#include <windows.h>

/**
 * Check if a directory exists.
 * @param path: The path of the directory to check.
 * @return: True if the directory exists, false otherwise.
 */
bool directoryExists(const std::wstring& path) {
    DWORD attributes = GetFileAttributesW(path.c_str());
    return (attributes != INVALID_FILE_ATTRIBUTES && (attributes & FILE_ATTRIBUTE_DIRECTORY));
}

/**
 * Generate a random bit sequence and save it to a file.
 * @param filepath: The path to the file where the random sequence will be saved.
 */
void generateRandomSequence(const std::wstring& filepath) {
    std::wstring directory = filepath.substr(0, filepath.find_last_of(L"/\\"));
    if (!directoryExists(directory)) {
        if (!CreateDirectoryW(directory.c_str(), NULL)) {
            std::cerr << "Failed to create directory\n";
            return;
        }
    }
    std::random_device rd;
    std::mt19937 generator(rd());
    std::ofstream outFile(filepath);
    if (!outFile) {
        std::cerr << "The file could not be opened for writing\n";
        return;
    }
    std::bitset<128> randomSequence;
    for (int j = 0; j < 128; ++j) {
        randomSequence[j] = generator() % 2; 
    }
    outFile << randomSequence;
    outFile.close();
}

int main() {
    std::wstring filepath = L"C:\\Users\\79297\\Desktop\\isb\\lab_2\\random_sequence_cc.txt";
    generateRandomSequence(filepath);
    
    return 0;
}

File: lab_5/generators/random_generator.java
import java.io.BufferedWriter;
import java.io.File;
import java.io.FileWriter;
import java.io.IOException;
import java.util.Random;

public class random_generator {
    /**
     * Generate a random bit sequence and save it to a file.
     * @param filename: The path to the file where the random sequence will be saved.
     */
    public static void main(String[] args) {
        String filename = "C:\\Users\\79297\\Desktop\\isb\\lab_2\\random_sequence_java.txt";
        try {
            File file = new File(filename);
            File directory = file.getParentFile();
            if (directory != null && !directory.exists()) {
                directory.mkdirs();
            }
            Random random = new Random();
            BufferedWriter writer = new BufferedWriter(new FileWriter(file));
            for (int i = 0; i < 128; ++i) {
                int bit = random.nextInt(2); 
                writer.write(Integer.toString(bit));
            }
            writer.close();
        } catch (IOException e) {
            System.err.println("The file could not be opened for writing");
            e.printStackTrace();
        }
    }
}
